orderer/common/cluster/clusterservice.go

```diff
@@ -0,0 +1,257 @@
+/*
+Copyright IBM Corp. 2017 All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package cluster
+
+import (
+	"bytes"
+	"encoding/asn1"
+	"fmt"
+	"io"
+	"strconv"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/hyperledger/fabric-protos-go/msp"
+	"github.com/hyperledger/fabric-protos-go/orderer"
+	"github.com/hyperledger/fabric/common/flogging"
+	"github.com/hyperledger/fabric/common/util"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+//go:generate mockery --dir . --name ClusterStepStream --case underscore --output ./mocks/
+
+// ClusterStepStream defines the gRPC stream for sending
+// transactions, and receiving corresponding responses
+type ClusterStepStream interface {
+	Send(response *orderer.ClusterNodeServiceStepResponse) error
+	Recv() (*orderer.ClusterNodeServiceStepRequest, error)
+	grpc.ServerStream
+}
+
+type AuthRequestVerifier interface {
+	VerifyAuthRequest(orderer.ClusterNodeService_StepServer, *orderer.ClusterNodeServiceStepRequest, TLSSessionBindingGetter, SignatureVerifier) (string, error)
+}
+
+type NodeIdentity struct {
+	ID       uint64
+	Identity []byte
+}
+
+// ClusterService defines the raft Service
+type ClusterService struct {
+	StreamCountReporter              *StreamCountReporter
+	Dispatcher                       Dispatcher
+	Logger                           *flogging.FabricLogger
+	StepLogger                       *flogging.FabricLogger
+	MinimumExpirationWarningInterval time.Duration
+	CertExpWarningThreshold          time.Duration
+	NodesCert                        map[string]map[uint64]*msp.SerializedIdentity
```

Comment:
1. Please add a lock that guards the NodesCert as you can have a reconfiguration from a different thread than a send.
2. Please call the NodesCert someting that resembles its use, like "membershipByChannel" or something like that

Reply:
added a lock to guard the map & renamed the map.